pep-0458.txt

```diff
@@ -504,22 +491,110 @@ MUST be offline and independent of other keys.  For simplicity of key
 management, without sacrificing security, it is RECOMMENDED that the keys of
 the *targets* role be permanently discarded as soon as they have been created
 and used to sign for the role.  Therefore, the *targets* role SHOULD require
-(1, 1) keys.  Again, this is because the keys are going to be permanently
+(2, 3) keys.  Again, this is because the keys are going to be permanently
 discarded and more offline keys will not help resist key recovery attacks [21]_
 unless diversity of keys is maintained.
 
+For similar reasons, the keys for the *bins* role SHOULD be set up similar to
+the keys for the *targets* role.
 
-Online and Offline Keys Recommended for Each Role
--------------------------------------------------
+In order to support continuous delivery, the keys for the *timestamp*,
+*snapshot*, and all *bin-n* roles MUST be online. There is little benefit in
+requiring all of these roles to use different online keys, since attackers
+would presumably be able to compromise all of them if they compromise PyPI.
+Therefore, it is reasonable to use one online key for them all.
 
-In order to support continuous delivery, the *timestamp*, *snapshot*, *bins*
-role keys MUST be online.
+
+Managing online keys
+---------------------
+
+The online key shared by the *timestamp*, *snapshot*, and all *bin-n* roles
+MAY be stored, encrypted or not, on the Python infrastructure. For example,
+the key MAY be kept on a self-hosted key management service (e.g. Hashicorp
+Vault__), or a third-party one (e.g. AWS KMS__, Google Cloud KMS__, or Azure
+Key Vault__).
+
+__ https://www.vaultproject.io/
+__ https://aws.amazon.com/kms/
+__ https://cloud.google.com/kms/
+__ https://docs.microsoft.com/en-us/azure/key-vault/basic-concepts
+
+Some of these key management services allow keys to be stored on Hardware
+Security Modules (HSMs) (e.g., Hashicorp Vault__, AWS CloudHSM__, Google
+Cloud HSM__, Azure Key Vault__). This prevents attackers from exfiltrating
+the online private key (albeit not from using it, although their actions
+may now be cryptographically auditable). However, this requires modifying
+the reference TUF implementation to support HSMs (WIP__).
+
+__ https://www.vaultproject.io/docs/enterprise/hsm/index.html
+__ https://aws.amazon.com/cloudhsm/
+__ https://cloud.google.com/hsm/
+__ https://docs.microsoft.com/en-us/azure/key-vault/key-vault-hsm-protected-keys
+__ https://github.com/secure-systems-lab/securesystemslib/pull/170
+
+Regardless of where and how this online key is kept, its use SHOULD be
+carefully logged, monitored, and audited.
```

Comment:
If so, separating keys may make these tasks easier (especially timestamp from snapshot from all bin keys).

Reply:
Maybe, but for little added security, just extra maintenance burden

Reply:
I guess this may be true.  It may or may not make auditing easier to have separate keys.  Would these keys be stored in hardware in any way or all software?

Reply:
Probably all software, if I had to guess. HSMs are not cheap to buy or use.

Reply:
Having said that, cloud providers have good logging and auditing capabilities these days, even if the key service is software-not-hardware-backed...